autoscaler/providers/google/provider.go

```diff
@@ -0,0 +1,130 @@
+package google
+
+import (
+	"context"
+	"fmt"
+	"github.com/laszlocph/woodpecker/autoscaler"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/cloudresourcemanager/v1"
+	compute "google.golang.org/api/compute/v1"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type provider struct {
+	project       string
+	projectId     string
+	zone          string
+	instanceGroup string
+	service       *compute.Service
+}
+
+type Option func(*provider)
+
+func WithProject(project string) Option {
+	return func(p *provider) {
+		p.project = project
+	}
+}
+
+func WithZone(zone string) Option {
+	return func(p *provider) {
+		p.zone = zone
+	}
+}
+
+func WithInstanceGroup(instanceGroup string) Option {
+	return func(p *provider) {
+		p.instanceGroup = instanceGroup
+	}
+}
+
+func New(opts ...Option) (autoscaler.Provider, error) {
+	p := new(provider)
+	for _, opt := range opts {
+		opt(p)
+	}
+	client, err := google.DefaultClient(context.Background(), compute.ComputeScope)
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to create gcp oauth2 client")
+		return nil, err
+	}
+
+	resourceManager, err := cloudresourcemanager.New(client)
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to create resource manager client")
+		return nil, err
+	}
+
+	project, err := resourceManager.Projects.Get(p.project).Do()
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to get project")
+		return nil, err
+	}
+	p.projectId = strconv.FormatInt(project.ProjectNumber, 10)
+
+	p.service, err = compute.New(client)
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to create compute client")
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (p provider) SetCapacity(n int, minimumAge time.Duration) error {
```

Comment:
Does this account for nodes that are running a build at the point of autoscaling? I fear nodes with running builds can be killed by the scaler.

Reply:
It may be best to leave this one unmerged or to have it marked as beta.

Yes, there is potential for a running node to be killed. This isn't great but the agents do appear to have graceful shutdown implement. However, GCP will kill an ondemand instance in 90 seconds or a preemptive one in 30 seconds so even with graceful shutdown they could still be forcibly deleted. 

In the environment I'm running this in everything is on preemptive instances so instances can disappear at any moment. The cost effectiveness of this outweights a developer sometimes being annoyed and having to hit rebuild.

Reply:
I have an autoscaler codebase unpublished, will try to merge the two. The mentioned problem is solved there, but this implementation also have merits.

Reply:
Oh interesting! if you push it to a branch I'll be sure to check it out.